pkg/cli/tail.go

```diff
@@ -222,16 +238,46 @@ func (c *TailCommand) Run(ctx context.Context, args []string) error {
 	return nil
 }
 
-func (c *TailCommand) tail(ctx context.Context) ([]*loggingpb.LogEntry, error) {
-	var p logPuller
-	if c.testPuller != nil {
-		p = c.testPuller
-	} else {
-		logClient, err := logging.NewClient(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create logging client: %w", err)
+func (c *TailCommand) streamTail(ctx context.Context) error {
+	var extra []validation.Validator
+	if c.flagAdditionalCheck {
+		extra = append(extra, validation.ValidateLabels)
+	}
+
+	logCh := make(chan *loggingpb.LogEntry)
+
+	var failCount int
+	var totalCount int
+
+	go func() {
+		for l := range logCh {
+			totalCount++
+			js, err := protojson.Marshal(l)
+			if err != nil {
+				c.Errf("failed to marshal log to json (InsertId: %q): %w", l.InsertId, err)
+				continue
+			}
+			c.Outf(stripSpaces(string(js)))
+			if c.flagValidate {
+				if err := validation.Validate(string(js), extra...); err != nil {
+					failCount++
+					c.Errf("failed to validate log (InsertId: %q): %w\n", l.InsertId, err)
+				} else {
+					c.Outf("Successfully validated log (InsertId: %q)\n", l.InsertId)
+				}
+				c.Outf("Validation failed for %d logs (out of %d)", failCount, totalCount)
+			}
 		}
-		p = cloudlogging.NewPuller(ctx, logClient, c.flagScope)
+	}()
+
+	return c.StreamTail(ctx, logCh)
```

Comment:
I don't see where you listen to the context cancel event here, something like 
```
	select {
	case <-ctx.Done():
		return fmt.Errorf("stream tail validate cancelled: %w", ctx.Err())
	case <-pullErr:
		return <-pullErr
	}
```

Reply:
Done